### PR TITLE
fix: don't duplicate advise for main -> test and main -> androidTest.

### LINKED ISSUE #1818: Duplicate add advice generated for transitive dependency across `implementation` and `testImplementation`
### Contributor Checklist
- [x] I have read the [Contributing guide](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/blob/main/CONTRIBUTING.asciidoc).
- [x] I have read the [Code of Conduct](https://github.com/autonomousapps/dependency-analysis-gradle-plugin/blob/main/CODE_OF_CONDUCT.md).
- [x] No part of this bug report was created with an LLM/AI.

### Plugin version
3.18.0 (latest)

### Gradle version
9.7.0 (latest)

### JDK version
25

### (Optional) Kotlin and Kotlin Gradle Plugin (KGP) version
2.4.10

### (Optional) Android Gradle Plugin (AGP) version
N/A

### (Optional) `reason` output for bugs relating to incorrect advice

```
------------------------------------------------------------
You asked about the dependency ':libraries:ListingModel'.
You have been advised to add this dependency to 'testImplementation'.
You have been advised to add this dependency to 'implementation'.
------------------------------------------------------------

Shortest path from :domains:Search:impl to :libraries:ListingModel for releaseCompileClasspath:
:domains:Search:impl
\--- :libraries:Listing
      \--- :libraries:ListingModel

Shortest path from :domains:Search:impl to :libraries:ListingModel for releaseRuntimeClasspath:
:domains:Search:impl
\--- :libraries:Listing
      \--- :libraries:ListingModel

Shortest path from :domains:Search:impl to :libraries:ListingModel for releaseUnitTestCompileClasspath:
:domains:Search:impl
\--- :libraries:Listing
      \--- :libraries:ListingModel

Shortest path from :domains:Search:impl to :libraries:ListingModel for releaseUnitTestRuntimeClasspath:
:domains:Search:impl
\--- :libraries:Listing
      \--- :libraries:ListingModel

Source: release, main
---------------------
* Uses 1 class: [...].listingmodel.ListingType (implies implementation).

Source: release, test
---------------------
* Uses 1 class: [...].listingmodel.ListingType (implies testImplementation).
* Imports 3 constants: [...].listingmodel.ListingType.CAROUSEL, [...].listingmodel.ListingType.LIST, [...].listingmodel.ListingType.MOSAIC (implies testImplementation).
```

### Describe the bug

Add advice is generated for `implementation` and `testImplementation` for a transitive dependency, while only `implementation` should be given.

Adding both of these advice, leads to a new contradicting advice (to remove the `testImplementation` because it is implied by the `implementation`):

```
------------------------------------------------------------
You asked about the dependency ':libraries:ListingModel'.
You have been advised to remove this dependency from 'testImplementation'.
------------------------------------------------------------

Shortest path from :domains:Search:impl to :libraries:ListingModel for releaseCompileClasspath:
:domains:Search:impl
\--- :libraries:ListingModel

Shortest path from :domains:Search:impl to :libraries:ListingModel for releaseRuntimeClasspath:
:domains:Search:impl
\--- :libraries:ListingModel

Shortest path from :domains:Search:impl to :libraries:ListingModel for releaseUnitTestCompileClasspath:
:domains:Search:impl
\--- :libraries:ListingModel

Shortest path from :domains:Search:impl to :libraries:ListingModel for releaseUnitTestRuntimeClasspath:
:domains:Search:impl
\--- :libraries:ListingModel

Source: release, main
---------------------
* Uses 1 class: [..].listingmodel.ListingType (implies implementation).

Source: release, test
---------------------
* Uses 1 class: [..].listingmodel.ListingType (implies testImplementation).
* Imports 3 constants: [..].listingmodel.ListingType.CAROUSEL, [..].listingmodel.ListingType.LIST, [..].listingmodel.ListingType.MOSAIC (implies testImplementation).
```

### To Reproduce
I'll create a reproducer on a fork project.

### Expected behavior
Only the add advice on `implementation` should be given

### PR DESCRIPTION
**Pull request by autonomousapps**
Resolves https://github.com/autonomousapps/dependency-analysis-gradle-plugin/issues/1818

TODO
- [x] Add test for main -> androidTest
- [x] follow-up on https://github.com/autonomousapps/dependency-analysis-gradle-plugin/pull/1888#discussion_r4008037453
- [ ] follow-up on https://github.com/autonomousapps/dependency-analysis-gradle-plugin/pull/1888#discussion_r4008291545

### CODE REVIEW

File: src/main/kotlin/com/autonomousapps/model/internal/declaration/Bucket.kt [hunk 1]
```diff
@@ -125,10 +125,25 @@ internal enum class Bucket(val value: String) {
...
+      }
+    }
+
+    private fun declarationMatches(
+      buckets: List<Bucket>,
+      declarations: Set<Declaration>,
+      configurationNames: ConfigurationNames,
+    ): Boolean {
+      return declarations.isEmpty() || declarations.any { declaration ->
+        buckets.any { bucket -> bucket.matches(declaration, configurationNames) }
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/model/internal/declaration/Bucket.kt]: This change makes sense to me but I don't know what if anything it might break. I think conceptually the code should return true if there are no declarations, but the kdoc is clear that there needs to be a main declaration. But that doesn't make sense. Not sure how it got to this state.

File: src/functionalTest/groovy/com/autonomousapps/android/projects/DuplicateAdviceProject.groovy [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  DuplicateAdviceProject(String agpVersion) {
+    super(agpVersion)
+    this.gradleProject = build()
+  }
+
+  private GradleProject build() {
+    return newAndroidGradleProjectBuilder()
+      .withAndroidLibProject('consumer') { s ->
+        s.sources = sources
+        s.manifest = null // TODO(tsr): fix defaults for android libs
```
Comment (trobalik-duolingo) [inline src/functionalTest/groovy/com/autonomousapps/android/projects/DuplicateAdviceProject.groovy:28]: follow up

File: src/main/kotlin/com/autonomousapps/model/internal/declaration/Bucket.kt [hunk 3]
```diff
@@ -74,12 +74,25 @@ internal enum class Bucket(val value: String) {
...
+     * TODO(tsr): this as a concept isn't really working. It's kind of like I'm trying to predict where the usage
+     *  analysis will ultimately go to prevent certain kinds of "redundant" advice. E.g., in typical projects, if
+     *  something is declared on `implementation`, it's already visible to `test` source (and `androidTest` source for
+     *  Android projects). (To say nothing about KMP.) In such a case, I don't want to tell users to re-declare on
+     *  `testImplementation` or `androidTestImplementation`. At the same time, if something is declared on
+     *  `compileOnly`, then it (and its graph) is *not* visible to `test` or `androidTest` source sets. This function is
+     *  trying to act as a little shim to handle these edge cases, when something more comprehensive and modeled is
+     *  probably called for.
+     */
+    fun determineVisibilityForTests(
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/model/internal/declaration/Bucket.kt:87]: Think about this.
